### PR TITLE
Mac-motivated improvements to build

### PR DESCRIPTION
**Pull request by mattdelco**
This change tried to make compiles on Mac go more smoothly, though to "test" to completely successfully I'd have to cut mixer, broker, and pilot (none of these pass on my Linux system either, partly due to dependencies on pilot/pkg/kube/config and kubernetes).

In Makefile I set up LOCAL_GOARCH and LOCAL_GOOS based on the currently running system, and then use these as defaults for GOARCH and GOOS.  The LOCAL_xxx is then passed to gobuild.sh when building in go/bin (to override its tendency to default to Linux/amd64 for unset vars).

I then got bin/linters.sh to gracefully handle uncustomized paths, and let it fix things it thought were wrong.

pkg/version/version_test.go seems to assume go1.10, so either we should bump up the requirement in the Makefile or change this test to handle the version of go that was used to compile.  For now I added an alternate string, with the intention that this permits the current go version and at least one prior.

make complains about a duplicated rule in tests/istio.mk , so I removed one.

### CODE REVIEW

**Comment by ldemailly** [comment]
so I can type "make" or "make test" (or "make e2e") on a mac and it will make the docker images needed and both the local native binaries and linux docker images without me having to change any env var ? if so 👍 💯 🎉 

/lgtm

**Comment by mattdelco** [comment]
"make" (which basically implies "make test") apparently depends on whether you've got k8s installed and configured (I thought the go unit tests were supposed to be standalone but clearly aren't)--in my original change I was tempted to cut mixer, pilot, and broker tests from "make test" on Mac but they don't pass on my Linux system either so I backed off on that.

As for docker, it at least won't get hung up at the generate_cert step--I don't have docker installed on my Mac, but if I turn all the "docker build" commands into "echo"s of same then "make docker" and "make GOOS=linux docker" run to completion on my Mac.

**Comment by ldemailly** [comment]
> I was tempted to cut mixer, pilot, and broker tests from "make test" on Mac

we have some work needed to move tests that aren't unit tests (that require either a real or simulated k8s env) from the real unit tests/self contained/parallelizable ones. but that doesn't mean we should start (afaik mixer tests are split between unit and integration etc)

> I don't have docker installed on my Mac

having docker installed isn't an unreasonable requirement, specially when working on istio's build :-)

**Comment by kyessenov** [comment]
I'm working to identify those tests and move them from unit tests. It's
fairly hard since we are not diligent in marking them properly.

**Comment by mattdelco** [comment]
Looks like the linter on Mac doesn't agree with the one in CircleCI, so I'll have to massage that part a bit more before this can be merged.

As for the failures I saw, here's what I initially saw (my lack of certain tools being installed is somewhat intentional, as it helps to make it more obvious when new dependencies are added):

<pre>
ok  	istio.io/istio/mixer/pkg/config	0.234s
2018-03-16T20:21:13.694968Z	info	Using user home k8s config - might affect real cluster ! Not found:  /Users/delco/istio/go/src/istio.io/istio/mixer/pkg/config/crd/../../../../pilot/pkg/kube/config
--- FAIL: TestGetAPIServerExtensionCACert (0.00s)
	admit_test.go:489: stat /Users/delco/.kube/config: no such file or directory
FAIL
FAIL	istio.io/istio/mixer/pkg/config/crd	1.264s

ok  	istio.io/istio/broker/pkg/model/osb	0.024s
2018-03-16T20:37:20.283155Z	info	using in-cluster configuration
2018-03-16T20:37:20.283203Z	warn	Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
2018-03-16T20:37:20.283214Z	warn	error creating inClusterConfig, falling back to default config: unable to load in-cluster configuration, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined
--- FAIL: TestStoreInvariant (0.00s)
	client_test.go:68: invalid configuration: no configuration has been provided
2018-03-16T20:37:20.283737Z	info	using in-cluster configuration
2018-03-16T20:37:20.283805Z	warn	Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
2018-03-16T20:37:20.283827Z	warn	error creating inClusterConfig, falling back to default config: unable to load in-cluster configuration, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined
--- FAIL: TestBrokerConfig (0.00s)
	client_test.go:68: invalid configuration: no configuration has been provided
FAIL

2018-03-16T22:11:53.207411Z	info	registering for apiVersion config.istio.io/v1alpha2
2018-03-16T22:11:53.207453Z	info	registering CRD "routerules.config.istio.io"
2018-03-16T22:11:53.208734Z	info	Failed to start in-process serverfailed to register custom resources. Post http://localhost:8080/apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions: dial tcp [::1]:8080: getsockopt: connection refused
FAIL	istio.io/istio/pilot/pkg/bootstrap	0.187s

2018-03-16T22:12:48.495247Z	info	AdmissionReview for /, Kind=: / response=&AdmissionResponse{UID:,Allowed:false,Result:&k8s_io_apimachinery_pkg_apis_meta_v1.Status{ListMeta:ListMeta{SelfLink:,ResourceVersion:,Continue:,},Status:Failure,Message:configuration is invalid: empty key,Reason:BadRequest,Details:nil,Code:400,},Patch:nil,PatchType:nil,}
2018-03-16T22:12:48.503449Z	info	Using user home k8s config - might affect real cluster ! Not found:  /Users/delco/istio/go/src/istio.io/istio/pilot/pkg/kube/admit/../config
--- FAIL: TestGetAPIServerExtensionCACert (0.00s)
	admit_test.go:456: kubernetes configuration file "/Users/delco/.kube/config" does not exist
FAIL
FAIL	istio.io/istio/pilot/pkg/kube/admit	0.199s

2018-03-16T22:13:18.795167Z	info	Using user home k8s config - might affect real cluster ! Not found:  /Users/delco/istio/go/src/istio.io/istio/pilot/pkg/serviceregistry/kube/config
--- FAIL: TestServices (0.00s)
	controller_test.go:39: kubernetes configuration file "/Users/delco/.kube/config" does not exist


</pre>

**Comment by ldemailly** [comment]
for linters we have to use an image, afaik, to get uniform results

File: pkg/version/version_test.go [hunk 1]
```diff
@@ -18,10 +18,11 @@ import "testing"
...
 	cases := []struct {
-		name     string
-		in       BuildInfo
-		want     string
-		longWant string
+		name        string
+		in          BuildInfo
+		want        string
+		longWant    string
+		longWantAlt string
```
Comment (ldemailly) [inline pkg/version/version_test.go]: What is this new field, where does it come from, is there a doc ?

Comment (mattdelco) [reply pkg/version/version_test.go]: Afaict the anonymous struct is only used within this function to compare the output of the function against a hard-coded string.  The string is hard-coded for go 1.10, so I made an alternate string that can be matched against to pass the test (in this case a hard-coded string with 1.9.3).  I suppose I could go a step further and have this alternate be a generated string that matches the go version.  I could just use this generated string without having an alternate but it might defeat the utility of the test (i.e., if the actual code and the test code are using the same string substitution mechanism then I don't think the test provides much value [while then also introducing a shadow implementation to keep consistent with the actual).

Comment (mattdelco) [reply pkg/version/version_test.go]: I came up with an alternate implementation that's dynamic yet sufficiently different in implementation to retain some value of the test.  I meant to put it on a different branch for comparison but I fat-fingered the push.  I think I like the use of runtime.Version() better than what istio is using in the actual code (i.e., -X with bin/get_workspace_status), though I probably need guidance on how to make it look more like go code.  The var at the top looks too much like C, but if I stick it in the struct it gets a bit tough to read.

File: tests/istio.mk [hunk 2]
```diff
@@ -180,22 +180,3 @@ test/minikube/auth/e2e_pilot_alpha1: istioctl
         --ns pilot-auth-system \
         -n pilot-test \
            ${TESTOPTS} | tee ${OUT_DIR}/tests/test-report-auth-pilot-v1.raw
-
```
Comment (costinm) [inline tests/istio.mk]: Why remove this target ? It is used in circle.

Comment (mattdelco) [reply tests/istio.mk]: There's two of them.  Any invocation of make complains:
<pre>
tests/istio.mk:186: warning: overriding recipe for target 'test/minikube/noauth/e2e_pilot_alpha1'
tests/istio.mk:128: warning: ignoring old recipe for target 'test/minikube/noauth/e2e_pilot_alpha1'
</pre>

**Comment by costinm** [comment]
AFAIK the plan is for developers to locally run both 'real unit tests' (golden files, no deps) and small tests - requiring an apiserver, but fast. 

I don't mind if they are in separate targets - but one way or another you will need an apiserver and etcd. 
Louis got the new docker for mac working - which apparently includes an apiserver. 

On Linux the tests rely on 'testEnvLocal..' to be run first - not sure how it got removed from the deps of 'test'. AFAIK the tests are passing if this is run first ( circle is calling make test after setting up the local env with same script).  I strongly suggest changing that script to point to Louis' docker env on Mac, so 
all tests (including those testing k8s registry ) are run.

**Comment by mattdelco** [comment]
AFAICT it looks like the execution of 'testEnvLocal...' was explicitly added to the various CIs.  I'm not sure if it got removed from the deps of 'test' so that the CIs have more flexibility in which env to start up first (but if so it seems like 'test' should have been kept simple & standalone for devs and CIs could run something else that skips the setup step).  Anyway, I tried to make testEnvLocalK8S.s have better support for Mac but I seem to have hit a roadblock that k8s doesn't seem to ship a compiled kube-apiserver for Mac.

**Comment by costinm** [comment]
On mac we need to start minikube or docker env if we can't compile
apiserver...

**Comment by ldemailly** [comment]
> In Makefile I set up LOCAL_GOARCH and LOCAL_GOOS based on the currently running system, and then use these as defaults for GOARCH and GOOS. The LOCAL_xxx is then passed to gobuild.sh when building in go/bin (to override its tendency to default to Linux/amd64 for unset vars).

shouldn't just be the other way around, default is native, except for docker where it must be linux ?
(ignoring the ibm390 docker images for a second)


> apiserver

that part should be optional until we have something turnkey

File: bin/testEnvLocalK8S.sh [hunk 3]
```diff
@@ -76,14 +78,24 @@ function getDeps() {
      elif [ -f /tmp/apiserver/kube-apiserver ] ; then
        ln -s /tmp/apiserver/kube-apiserver $TOP/bin/
      else
-       curl -Lo ${TOP}/bin/kube-apiserver https://storage.googleapis.com/kubernetes-release/release/${K8S_VER}/bin/linux/amd64/kube-apiserver && chmod +x ${TOP}/bin/kube-apiserver
+       # bucket doesn't contain a kube-apiserver for darwin
+       curl -Lo ${TOP}/bin/kube-apiserver https://storage.googleapis.com/kubernetes-release/release/${K8S_VER}/bin/${LOCAL_GOOS}/amd64/kube-apiserver && chmod +x ${TOP}/bin/kube-apiserver
```
Comment (ldemailly) [inline bin/testEnvLocalK8S.sh]: eventually we need to nuke those multi curls across the code base (we already removed the ones in init.sh, replaced by vendor like they should have been from day 1) and have all the dependencies in 1 single vendor like download or image dependency

File: tests/istio.mk (hunk 2 above)
Comment (ldemailly) [reply tests/istio.mk]: thanks for fixing this @mattdelco

File: bin/testEnvLocalK8S.sh (hunk 3 above)
Comment (mattdelco) [reply bin/testEnvLocalK8S.sh]: I agree it's be nice (I'd tend towards also porting the file to make), though I'm tending towards smaller PRs that are more likely for me to get to completion now that I've got another day job (despite my other change that I'll probably turn into a PR once vendor has the newer fortio).  I did try to run a "make test" on my Linux box and first ran the make target that runs this script, but I still got failures (mostly from a non-present kubeconfig file, and creating an empty one just changes the complain message but the overall result is still a failure).

Comment (ldemailly) [reply bin/testEnvLocalK8S.sh]: a (somewhat low pri) test of our makefile infra is if can run from "docker run ubuntu:xenial" (+ git + go + docker - I have such an image actually it's `istio/fortio.build:v6` but we should make one specifically for istio "install from scratch is possible" tests )

**Comment by mattdelco** [comment]
> shouldn't just be the other way around, default is native, except for docker where it
> must be linux (ignoring the ibm390 docker images for a second)

I'm not sure if I just didn't state things clearly.  For the most part things do default to native.  The main part I changed affects when creating build tools in bin/ .  These tools need to be compiled for native regardless of any cross-compile request.  Previously this was done by trying to unset GOOS/GOARCH when calling gobuild.sh, but that script defaults unset to Linux/amd64 so rather than unset I explicitly set GOOS/GOARCH.  When using make the defaults in gobuild.sh normally don't come into play because the Makefile has already set GOOS/GOARCH with value that default to native.

As for docker, I don't know if the eventual goal is for Mac compiles to 1) create Mac-native docker images (ignoring the debug images that use a ubuntu base, as the rest are mostly "from scratch" and I presume have some chance of working) or 2) default to cross-compiling docker images for Linux.

If the goal is the latter (i.e., default to Linux docker images when building on mac), then the Makefile should probably look something like:

<pre>
ifeq ($(GOOS),linux)
include tools/istio-docker.mk
else
docker:
        $(MAKE) GOOS=linux GOARCH=amd64 docker
endif
</pre>

File: bin/testEnvLocalK8S.sh (hunk 3 above)
Comment (mattdelco) [reply bin/testEnvLocalK8S.sh]: It'll probably also need fpm, at least if you care about the debian.  Costin has previously advocated for using the image that's used by CircleCI (the debian makefile has an optional target that tries to build the debian file from within a docker image).  There's yet another image used by prow/Cloud Builder.  So, everybody has their favorite. I've tried to stay out of the image debates and focused on getting the Makefile to check/resolve its own dependencies that can be conveniently run locally (basically excludes git, go, docker, and probably fpm too),  or at least have it check and complain when something's missing.  I just don't know yet what's actually needed to get "make test" to pass (either in theory or actuality).

Comment (ldemailly) [reply bin/testEnvLocalK8S.sh]: sgtm

**Comment by ldemailly** [comment]
I guess my point was that you don't need any variable set at all to do the default go behavior and only when making docker image do we need to use linux (and static etc...)
it's back to why is there even a gobuild.sh; there should only be a "dockerbuild.sh" or even better build docker images using docker (like https://github.com/istio/fortio/blob/master/Dockerfile#L36)

**Comment by mattdelco** [comment]
I think bin/gobuild.sh mostly exists to emulate Bazel's get_workspace_status behavior.  Both bin/gobuild.sh and bin/get_workspace_status can be ported to Make (ideally with something like the "BUILDINFO" caching behavior that afaik no build/test system has yet adopted).  The default to Linux/amd64 in gobuild.sh might be a reflection of some earlier philosophies I've heard expressed that the Makefiles were mostly for release builds/tests (i.e., Linux), and go devs would just use native tools (though I'm not clear on how go tools would have picked up the get_workspace_status functionality).

When native is desired I only explicitly set GOOS/GOARCH when calling gobuild.sh.  For other cases (e.g., "go get" or "go install") these parameters are unset (this is probably more evident in the other PR I just made since it downloads more tools).

**Comment by ldemailly** [comment]
i'd strongly prefer to head in the direction of "make" doing the right thing, without having to workaround bad defaults inside gobuild.sh; I feel like a broken record but afaik _only_ Docker needs to set any env, and you don't even if we were building multi-stage builds like we should be.

**Comment by mattdelco** [comment]
> i'd strongly prefer to head in the direction of "make" doing the right thing, without having to workaround
> bad defaults inside gobuild.sh; I feel like a broken record but afaik only Docker images needs to set
> any env, and would not even if we were building multi-stage builds like we should be.

I'm not clear if "make doing the right thing" means change gobuild.sh to not default to linux/amd64 (and thus the Makefile wouldn't need to override this).  If so, then I'd have in mind to do this separately (possibly by getting rid of gobuild.sh).  Otherwise, you'll need to be more explicit about what you had in mind.  I think with this current change the invocations of make basically fit people's expectations (i.e. make compiles for the native OS unless you opt-in to cross-compiles), even though the internal means of carrying things out are a bit convoluted.

> So they do have kubeapiserver builds for mac ? I didn't know this - it simplifies things.

The apiserver isn't there yet, at least not in the expected bucket (I examined the bucket contents to make sure it wasn't just a matter of change in file name/type convention -- from what I saw the Mac bucket only has a few files in it).  I did change part of the kube-apiserver URL from "linux" to "${LOCAL_GOOS}" for the sake of completeness and consistency, but that just means that on Mac bin/testEnvLocalK8S.sh will now fail at trying to download a non-existent file, rather than downloading a Linux binary and failing when "kill -0 $(cat $LOG_DIR/apiserver.pid)" can't find a running process. To help avoid confusion I added a "bucket doesn't contain a kube-apiserver for darwin" comment (so someone doesn't think this did work at some point in the past).  Perhaps I should've gone a step further and just have the script explicitly fail with msg on Mac, but I wasn't sure if some people were going so far as to compile K8s for their Mac.

**Comment by ldemailly** [comment]
I guess it's a good first step if it just works on a mac; we can fix _how_ it ends up working later (ie getting rid of the env vars/ gobuild.sh etc) - whichever you can do - thanks !

**Comment by mattdelco** [comment]
After the rebase it looks like another PR opted for the name "GOOS_LOCAL" over "LOCAL_GOOS" so I renamed the vars in this PR to match (and renamed the new "LOCAL_GOARCH" to "GOARCH_LOCAL" to stay consistent, even though it seems odd to me to have some vars called LOCAL_xxx and others called xxx_LOCAL).

I'd request a retest on the 2 failures but it looks like nobody's managed to pass them in the last ~6 hours.
